File: android/src/main/java/com/mycompany/capacitor/device/check/CapacitorDeviceCheckPlugin.kt
package com.mycompany.capacitor.device.check

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "CapacitorDeviceCheck")
public class CapacitorDeviceCheckPlugin : Plugin() {
    private val implementation = CapacitorDeviceCheck()

    @PluginMethod
    public fun generateToken(call: PluginCall) {
        val ret = JSObject()
        ret.put("token", implementation.generateToken())
        call.resolve(ret)
    }
}

File: ios/Sources/CapacitorDeviceCheckPlugin/CapacitorDeviceCheckPlugin.swift
import Foundation
import Capacitor
import DeviceCheck

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorDeviceCheckPlugin)
public class CapacitorDeviceCheckPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorDeviceCheckPlugin"
    public let jsName = "CapacitorDeviceCheck"
    public let pluginMethods: [CAPPluginMethod] = [
        .async("generateToken", CapacitorDeviceCheckPlugin.generateToken)
    ]

    // generateToken touches no UIKit state, so it does not need the main actor.
    func generateToken(_ call: CAPPluginCall) async throws -> JSObject {
        guard DCDevice.current.isSupported == true else {
            throw CAPPluginError("DeviceCheck is not supported on this device")
        }
        let deviceToken: String = try await withCheckedThrowingContinuation { continuation in
            // DeviceCheck calls the handler once.
            DCDevice.current.generateToken { tokenData, error in
                if let error {
                    continuation.resume(throwing: CAPPluginError("DeviceCheck error:\(error.localizedDescription)", underlyingError: error))
                } else if let deviceToken = tokenData?.base64EncodedString() {
                    continuation.resume(returning: deviceToken)
                } else {
                    continuation.resume(throwing: CAPPluginError("DeviceCheck token encoding failed"))
                }
            }
        }
        return ["token": deviceToken]
    }
}
